README.md
# hexagonal_grid_widget

A hex layout widget that supports infinite omniscroll and is highly customizable. Only the hex 
widgets that are viewable are rendered while all others are not, thus making this widget very 
efficient even at larger datasets.

# To Publish to Pub
* Follow https://www.dartlang.org/tools/pub/publishing

--- or ---

* Make updates to code
* Update `pubspec.yaml` to the desired version number
* Update `CHANGELOG.md` with the version number and changes
* `pub publish --dry-run`
* `pub publish`


CHANGELOG.md
## [1.0.3] - 12/10/2019

* Implemented ability to choose flat or pointy hex layouts. Defaulting to flat to preserve original functionality.

## [1.0.2] - 03/01/2019

* Fixed a bug that did not allow for the hex grid children to update properly on state change

## [1.0.0] - 01/17/2019

* Packagized this hex grid widget that I have been using in my own projects so other projects and
packages can use it. Provides a highly customizable and performance conscious widget to render 
a list of objects marshaled into a widget as a hex layout.


test/test_utils.dart
import 'dart:math';

import 'package:flutter/widgets.dart';
import 'package:hexagonal_grid/hexagonal_grid.dart';
import 'package:hexagonal_grid_widget/hex_grid_child.dart';
import 'package:hexagonal_grid_widget/hex_grid_context.dart';

class TestUtils {
  static List<HexGridChild> createHexGridChildren(int numOfChildren) {
    List<HexGridChild> children = [];

    for (int i = 0; i < numOfChildren; i++) {
      children.add(TestHexGridChild(i));
    }

    return children;
  }
}

class TestHexGridChild extends HexGridChild {
  final int index;
  final List<Color> orbitalColors = [
    Color(0xFF2D365C),
    Color(0xFF083663),
    Color(0xFF07489C),
    Color(0xFF165DC0),
    Color(0xFF0E90E1),
    Color(0xFF89D3FB),
    Color(0xFFAFDBDE)
  ];

  TestHexGridChild(this.index);

  @override
  Widget toHexWidget(BuildContext context, HexGridContext hexGridContext,
      double size, UIHex hex) {
    return Container(
        padding: EdgeInsets.all((hexGridContext.maxSize - size) / 2),
        child: Container(
            width: size,
            height: size,
            decoration: BoxDecoration(
              color: orbitalColors[hex.orbital % orbitalColors.length],
              shape: BoxShape.circle,
            )));
  }

  @override
  double getScaledSize(
      HexGridContext hexGridContext, double distanceFromOrigin) {
    double scaledSize = hexGridContext.maxSize -
        (distanceFromOrigin * hexGridContext.scaleFactor);
    return max(scaledSize, hexGridContext.minSize);
  }
}


test/hex_grid_widget_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:hexagonal_grid_widget/hex_grid_child.dart';
import 'package:hexagonal_grid_widget/hex_grid_context.dart';
import 'package:hexagonal_grid_widget/hex_grid_widget.dart';

import 'test_utils.dart';

void main() {
  final double _minHexWidgetSize = 24;
  final double _maxHexWidgetSize = 128;
  final double _scaleFactor = 0.2;
  final double _densityFactor = 1.75;
  final double _velocityFactor = 0.3;

  testWidgets('10 HexGridChild widgets are visible and are in the render tree',
      (WidgetTester tester) async {
    final int numOfChildren = 10;
    final List<HexGridChild> children =
        TestUtils.createHexGridChildren(numOfChildren);

    final HexGridWidget hexGridWidget = HexGridWidget(
        children: children,
        hexGridContext: HexGridContext(_minHexWidgetSize, _maxHexWidgetSize,
            _scaleFactor, _densityFactor, _velocityFactor));

    await tester.pumpWidget(MaterialApp(
      home: Scaffold(body: hexGridWidget),
    ));

    //The HexGridWidget needs to find the origin of its container, thus it must
    // wait at least one frame after rendering to determine the width and height
    // to determine its center. Because of this we must wait to the layout
    // renderer to "settle" or fully complete its lifecycle changes.
    await tester.pumpAndSettle();

    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.byType(Positioned)),
        findsNWidgets(numOfChildren));
    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.widgetWithText(RaisedButton, 'Center')),
        findsNothing);
  });

  testWidgets(
      '10 HexGridChild and a Center widget are rendered when the '
      'omniscroll is off center', (WidgetTester tester) async {
    final int numOfChildren = 10;
    final List<HexGridChild> children =
        TestUtils.createHexGridChildren(numOfChildren);

    final HexGridWidget hexGridWidget = HexGridWidget(
        children: children,
        hexGridContext: HexGridContext(_minHexWidgetSize, _maxHexWidgetSize,
            _scaleFactor, _densityFactor, _velocityFactor));

    await tester.pumpWidget(MaterialApp(
      home: Scaffold(body: hexGridWidget),
    ));

    //Wait for the state to catch up
    await tester.pumpAndSettle();

    //Find the 10 HexGridChild widgets, but no Center text widget
    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.byType(Positioned)),
        findsNWidgets(numOfChildren));
    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.widgetWithText(RaisedButton, 'Center')),
        findsNothing);

    //Artificially scroll (Set offset) to the top-left by maxHexWidgetSize
    Offset center = tester.getCenter(find.descendant(
        of: find.byWidget(hexGridWidget),
        matching: find.byType(GestureDetector)));
    hexGridWidget.offset =
        center + Offset(_maxHexWidgetSize, _maxHexWidgetSize);

    //Wait for the state to catch up
    await tester.pumpAndSettle();

    //Now all 10 HexGridChild widgets should still show as well as a Center
    // text widget
    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.byType(Positioned)),
        findsNWidgets(numOfChildren));
    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.widgetWithText(RaisedButton, 'Center')),
        findsOneWidget);
  });

  testWidgets('Only HexGridChild widgets that are visiable are rendered',
      (WidgetTester tester) async {
    final int numOfChildren = 10;
    final List<HexGridChild> children =
        TestUtils.createHexGridChildren(numOfChildren);

    final HexGridWidget hexGridWidget = HexGridWidget(
        children: children,
        hexGridContext: HexGridContext(_minHexWidgetSize, _maxHexWidgetSize,
            _scaleFactor, _densityFactor, _velocityFactor));

    await tester.pumpWidget(MaterialApp(
      home: Scaffold(body: hexGridWidget),
    ));

    //Wait for the state to catch up
    await tester.pumpAndSettle();

    //All 10 HexGridChild widgets should show, but no Center text widget
    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.byType(Positioned)),
        findsNWidgets(numOfChildren));
    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.widgetWithText(RaisedButton, 'Center')),
        findsNothing);

    //Artificially scroll (Set offset) to the top-left by doubling center
    Offset center = tester.getCenter(find.descendant(
        of: find.byWidget(hexGridWidget),
        matching: find.byType(GestureDetector)));
    hexGridWidget.offset = center + center;

    //Wait for the state to catch up
    await tester.pumpAndSettle();

    //Now only 7 HexGridChild widgets are in view, thus only 7 should be in
    // the render tree. The Center text widget should also show
    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.byType(Positioned)),
        findsNWidgets(7));
    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.widgetWithText(RaisedButton, 'Center')),
        findsOneWidget);

    //Artificially scroll (Set offset) back to the center
    hexGridWidget.offset = center;

    //Wait for the state to catch up
    await tester.pumpAndSettle();

    //All 10 HexGridChild widgets are in view and thus all 10 should be in
    // the render tree. The Center text widget should no longer show
    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.byType(Positioned)),
        findsNWidgets(numOfChildren));
    expect(
        find.descendant(
            of: find.byWidget(hexGridWidget),
            matching: find.widgetWithText(RaisedButton, 'Center')),
        findsNothing);
  });
}


lib/hex_grid_widget.dart
library hexagonal_grid_widget;

import 'dart:math';

import 'package:after_layout/after_layout.dart';
import 'package:flutter/material.dart';
import 'package:hexagonal_grid/hexagonal_grid.dart';
import 'package:hexagonal_grid_widget/hex_grid_child.dart';
import 'package:hexagonal_grid_widget/hex_grid_context.dart';
import 'package:tuple/tuple.dart';

@immutable
class HexGridWidget<T extends HexGridChild> extends StatefulWidget {
  HexGridWidget(
      {@required this.hexGridContext,
      @required this.children,
      this.scrollListener});

  final HexGridContext hexGridContext;
  final List<T> children;

  final ValueChanged<Offset> scrollListener;
  final ValueNotifier<Offset> offsetNotifier = ValueNotifier(Offset(0, 0));

  @override
  State<StatefulWidget> createState() =>
      _HexGridWidgetState(hexGridContext, scrollListener, offsetNotifier);

  //Set the x and y scroll offset
  set offset(Offset offset) {
    offsetNotifier.value = offset;
  }
}

// ignore: conflicting_generic_interfaces
class _HexGridWidgetState<T extends HexGridChild> extends State<HexGridWidget>
    with SingleTickerProviderStateMixin, AfterLayoutMixin<HexGridWidget> {
  final GlobalKey _containerKey = GlobalKey();
  bool _isAfterFirstLayout = false;

  HexGridContext _hexGridContext;
  List<UIHex> _hexLayout;
  double _hexLayoutRadius = 0.0;

  double xPos = 0.0;
  double yPos = 0.0;
  double xViewPos = 0.0;
  double yViewPos = 0.0;
  Point origin = Point(0.0, 0.0);

  Animation<Offset> _flingAnimation;
  bool _enableFling = false;

  AnimationController _controller;
  ValueChanged<Offset> _scrollListener;
  ValueNotifier<Offset> _offsetNotifier;

  _HexGridWidgetState(
      HexGridContext hexGridContext,
      ValueChanged<Offset> scrollListener,
      ValueNotifier<Offset> offsetNotifier) {
    _hexGridContext = hexGridContext;

    if (scrollListener != null) {
      _scrollListener = scrollListener;
    }

    if (offsetNotifier != null) {
      _offsetNotifier = offsetNotifier;
      _offsetNotifier.addListener(updateOffsetFromNotifier);
    }
  }

  @override
  void initState() {
    super.initState();

    _hexLayout = UIHex.toSpiralHexLayout(widget.children);
    _isAfterFirstLayout = false;

    _controller = AnimationController(vsync: this)
      ..addListener(_handleFlingAnimation);
  }

  @override
  void dispose() {
    _controller?.dispose();

    //Don't dispose as other's might be using it. It would be up to the owner,
    // in this case HexGridWidget, to dispose of it. So only clean up after
    // ourselves (this class, _HexGridWidgetState)
    _offsetNotifier?.removeListener(updateOffsetFromNotifier);

    super.dispose();
  }

  @override
  void afterFirstLayout(BuildContext context) {
    _isAfterFirstLayout = true;

    final double containerWidth = this.containerWidth;
    final double containerHeight = this.containerHeight;

    //Determine the origin of the container. Since we'll be using origin w.r.t
    // to the bounding boxes of the hex children, which are positioned by
    // top and left values, we'll have to adjust by half of the widget size to
    // get the technical origin.
    origin = Point((containerWidth / 2) - (_hexGridContext.maxSize / 2),
        (containerHeight / 2) - (_hexGridContext.maxSize / 2));

    //Center the hex grid to origin
    offset = Offset(origin.x, origin.y);
  }

  void updateOffsetFromNotifier() => offset = _offsetNotifier.value;

  set offset(Offset offset) {
    setState(() {
      xViewPos = offset.dx;
      yViewPos = offset.dy;
    });
  }

  double get containerHeight {
    RenderBox containerBox = _containerKey.currentContext.findRenderObject();
    return containerBox.size.height;
  }

  double get containerWidth {
    RenderBox containerBox = _containerKey.currentContext.findRenderObject();
    return containerBox.size.width;
  }

  ///Ensures we will always have hex widgets visible
  Tuple2<double, double> _confineHexGridWithinContainer(
      double newXPosition, double newYPosition) {
    //Don't allow the right of the hex grid widget to exceed pass the left half
    // of the container
    if (newXPosition < origin.x - _hexLayoutRadius) {
      newXPosition = origin.x - _hexLayoutRadius;
    }

    //Don't allow the left of the hex grid widget to exceed pass the right half
    // of the container
    if (newXPosition > origin.x + _hexLayoutRadius) {
      newXPosition = origin.x + _hexLayoutRadius;
    }

    //Don't allow the bottom of the hex grid widget to exceed pass the top half
    // of the container
    if (newYPosition < origin.y - _hexLayoutRadius) {
      newYPosition = origin.y - _hexLayoutRadius;
    }

    //Don't allow the top of the hex grid widget to exceed pass the bottom half
    // of the container
    if (newYPosition > origin.y + _hexLayoutRadius) {
      newYPosition = origin.y + _hexLayoutRadius;
    }

    return Tuple2<double, double>(newXPosition, newYPosition);
  }

  void _handleFlingAnimation() {
    if (!_enableFling ||
        _flingAnimation.value.dx.isNaN ||
        _flingAnimation.value.dy.isNaN) {
      return;
    }

    double newXPosition = xPos + _flingAnimation.value.dx;
    double newYPosition = yPos + _flingAnimation.value.dy;

    Tuple2<double, double> newPositions =
        _confineHexGridWithinContainer(newXPosition, newYPosition);

    newXPosition = newPositions.item1;
    newYPosition = newPositions.item2;

    setState(() {
      xViewPos = newXPosition;
      yViewPos = newYPosition;
    });

    _sendScrollValues();
  }

  void _handlePanUpdate(DragUpdateDetails details) {
    final RenderBox referenceBox = context.findRenderObject();
    final Offset position = referenceBox.globalToLocal(details.globalPosition);

    double newXPosition = xViewPos + (position.dx - xPos);
    double newYPosition = yViewPos + (position.dy - yPos);

    Tuple2<double, double> newPositions =
        _confineHexGridWithinContainer(newXPosition, newYPosition);

    newXPosition = newPositions.item1;
    newYPosition = newPositions.item2;

    setState(() {
      xViewPos = newXPosition;
      yViewPos = newYPosition;
    });

    xPos = position.dx;
    yPos = position.dy;

    _sendScrollValues();
  }

  void _handlePanDown(DragDownDetails details) {
    _enableFling = false;
    final RenderBox referenceBox = context.findRenderObject();
    final Offset position = referenceBox.globalToLocal(details.globalPosition);

    xPos = position.dx;
    yPos = position.dy;
  }

  void _handlePanEnd(DragEndDetails details) {
    final double magnitude = details.velocity.pixelsPerSecond.distance;
    final double velocity = magnitude / 1000;

    final Offset direction = details.velocity.pixelsPerSecond / magnitude;
    final double distance = (Offset.zero & context.size).shortestSide;

    xPos = xViewPos;
    yPos = yViewPos;

    _enableFling = true;
    _flingAnimation = Tween<Offset>(
            begin: Offset(0.0, 0.0),
            end: direction * distance * _hexGridContext.velocityFactor)
        .animate(_controller);
    _controller
      ..value = 0.0
      ..fling(velocity: velocity);
  }

  _sendScrollValues() {
    if (_scrollListener != null) {
      _scrollListener(Offset(xViewPos, yViewPos));
    }
  }

  @override
  Widget build(BuildContext context) {
    Widget childToShow;
    if (!_isAfterFirstLayout) {
      childToShow = Container();
    } else {
      childToShow = Stack(
          children: _buildHexWidgets(
              _hexGridContext.maxSize / _hexGridContext.densityFactor,
              xViewPos,
              yViewPos));
    }

    return GestureDetector(
      onPanDown: _handlePanDown,
      onPanUpdate: _handlePanUpdate,
      onPanEnd: _handlePanEnd,
      child: Container(
          decoration: BoxDecoration(
            color: Theme.of(context).backgroundColor,
          ),
          key: _containerKey,
          child: childToShow),
    );
  }

  List<Widget> _buildHexWidgets(
      double hexSize, double layoutOriginX, double layoutOriginY) {

    HexLayout layout;
    if (_hexGridContext.flatLayout) {
      layout = HexLayout.orientFlat(
          Point(hexSize, hexSize), Point(layoutOriginY, layoutOriginX));
    } else {
      layout = HexLayout.orientPointy(
          Point(hexSize, hexSize), Point(layoutOriginY, layoutOriginX));
    }

    List<Widget> hexWidgetList = [];

    final double containerWidth = this.containerWidth;
    final double containerHeight = this.containerHeight;

    for (int i = 0; i < _hexLayout.length; i++) {
      Positioned hexWidget = _createPositionWidgetForHex(widget.children[i],
          _hexLayout[i], layout, containerWidth, containerHeight);

      if (hexWidget != null) {
        hexWidgetList.add(hexWidget);
      }
    }

    if (_hexLayout.isNotEmpty) {
      final Point originHexToPixel = _hexLayout.first.hex.toPixel(layout);

      _hexLayoutRadius =
          originHexToPixel.distanceTo(_hexLayout.last.hex.toPixel(layout));

      if (originHexToPixel.y > origin.x + _hexGridContext.maxSize / 2 ||
          originHexToPixel.y < origin.x - _hexGridContext.maxSize / 2 ||
          originHexToPixel.x > origin.y + _hexGridContext.maxSize / 2 ||
          originHexToPixel.x < origin.y - _hexGridContext.maxSize / 2) {
        hexWidgetList.add(Align(
            alignment: Alignment.bottomCenter,
            child: Padding(
                padding: EdgeInsets.only(bottom: 16),
                child: RaisedButton(
                  child: Text("Center"),
                  elevation: 4,
                  textTheme: ButtonTextTheme.normal,
                  onPressed: () => _centerHexLayout(),
                ))));
      }
    }

    return hexWidgetList;
  }

  ///Only return a [Positioned] if the widget will be visible, otherwise return
  /// null so we don't waste CPU cycles on rendering something that's not visible
  /// NOTE: As with the rest of a Hex grid, the x and y coordinates are reflected
  Positioned _createPositionWidgetForHex(T hexGridChild, UIHex uiHex,
      HexLayout hexLayout, double containerWidth, double containerHeight) {
    final Point hexToPixel = uiHex.hex.toPixel(hexLayout);

    //If the right of the hex exceeds pass the left border of the container
    if (hexToPixel.y + _hexGridContext.maxSize < 0) {
      return null;
    }

    //If the left of the hex exceeds pass the right border of the container
    if (hexToPixel.y - _hexGridContext.maxSize > containerWidth) {
      return null;
    }

    //If the bottom of the hex exceeds pass the top border of the container
    if (hexToPixel.x + _hexGridContext.maxSize < 0) {
      return null;
    }

    //If the top of the hex exceeds pass the bottom border of the container
    if (hexToPixel.x - _hexGridContext.maxSize > containerHeight) {
      return null;
    }

    final Point reflectedOrigin = Point(origin.y, origin.x);
    final double distance = hexToPixel.distanceTo(reflectedOrigin);
    final double size = hexGridChild.getScaledSize(_hexGridContext, distance);

    return Positioned(
        top: hexToPixel.x,
        left: hexToPixel.y,
        child: hexGridChild.toHexWidget(context, _hexGridContext, size, uiHex));
  }

  void _centerHexLayout() {
    xPos = xViewPos;
    yPos = yViewPos;

    _enableFling = true;
    _flingAnimation = Tween<Offset>(
            begin: Offset(0, 0), end: Offset(origin.x - xPos, origin.y - yPos))
        .animate(_controller);
    _controller
      ..value = 0.0
      ..fling(velocity: 1);
  }
}


lib/hex_grid_context.dart
class HexGridContext {
  ///Minimum size of a individual hex widget
  final double minSize;

  ///Maximum size of a individual hex widget
  final double maxSize;

  ///Controls how significantly the hex child widgets shrink as the move further
  /// from the origin. The larger the number, the quicker the widgets will reduce
  /// in size as they get further from the origin
  final double scaleFactor;

  ///Controls how close the widgets sit next to each other. Note if the
  /// densityFactor is greater than two then the hex child widgets will overlap
  /// The larger the number the more dense the hex child widget will sit
  final double densityFactor;

  ///Controls the speed of the flingAnimation. The larger the number the faster
  /// the fling animation will play
  final double velocityFactor;

  ///Controls weather the geometry of the layout is pointy or flat. Defaults
  /// to flat
  final bool flatLayout;

  HexGridContext(this.minSize, this.maxSize, this.scaleFactor,
      this.densityFactor, this.velocityFactor, [this.flatLayout = true]);
}


example/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:hexagonal_grid/hexagonal_grid.dart';
import 'package:hexagonal_grid_widget/hex_grid_child.dart';
import 'package:hexagonal_grid_widget/hex_grid_context.dart';
import 'package:hexagonal_grid_widget/hex_grid_widget.dart';

void main() => runApp(HexGridWidgetExample());

class HexGridWidgetExample extends StatelessWidget {
  final double _minHexWidgetSize = 24;
  final double _maxHexWidgetSize = 128;
  final double _scaleFactor = 0.2;
  final double _densityFactor = 1.75;
  final double _velocityFactor = 0.3;
  final bool _flatLayout = true;
  final int _numOfHexGridChildWidgets = 10;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text("Example"),
          centerTitle: true,
        ),
        body: HexGridWidget(
            children: createHexGridChildren(_numOfHexGridChildWidgets),
            hexGridContext: HexGridContext(_minHexWidgetSize, _maxHexWidgetSize,
                _scaleFactor, _densityFactor, _velocityFactor, _flatLayout)));
  }

  //This would likely be a service (RESTful or DB) that retrieves some data and
  // marshals them into HexGridChild objects
  List<HexGridChild> createHexGridChildren(int numOfChildren) {
    List<HexGridChild> children = [];

    for (int i = 0; i < numOfChildren; i++) {
      children.add(ExampleHexGridChild(i));
    }

    return children;
  }
}

//This class can contain all the properties you'd like, but it must extends
// HexGridChild and thus implement the toHexWidget and getScaledSized methods.
// The methods provide most fields the HexGridWidget is aware of to allow for
// as much flexibility when building and sizing your HexGridChild widget.
class ExampleHexGridChild extends HexGridChild {
  final int index;
  final List<Color> orbitalColors = [
    Color(0xFF2D365C),
    Color(0xFF083663),
    Color(0xFF07489C),
    Color(0xFF165DC0),
    Color(0xFF0E90E1),
    Color(0xFF89D3FB),
    Color(0xFFAFDBDE)
  ];

  ExampleHexGridChild(this.index);

  //This is only one example of the customization you can expect from these
  // framework hooks
  @override
  Widget toHexWidget(BuildContext context, HexGridContext hexGridContext,
      double size, UIHex hex) {
    return Container(
        padding: EdgeInsets.all((hexGridContext.maxSize - size) / 2),
        child: Container(
            width: size,
            height: size,
            decoration: BoxDecoration(
              color: orbitalColors[hex.orbital % orbitalColors.length],
              shape: BoxShape.circle,
            )));
  }

  @override
  double getScaledSize(
      HexGridContext hexGridContext, double distanceFromOrigin) {
    double scaledSize = hexGridContext.maxSize -
        (distanceFromOrigin * hexGridContext.scaleFactor);
    return max(scaledSize, hexGridContext.minSize);
  }
}
